pkg/agent/openflow/client_test.go

```diff
@@ -152,58 +150,41 @@ func TestConcurrentFlowInstallation(t *testing.T) {
 			client := ofClient.(*client)
 			client.flowOperations = m
 
-			var wg sync.WaitGroup
-			var counter int32        // tracks the number of concurrent Add calls at any given time
-			var callIdx int32        // used to distinguish between the first Add call and subsequent calls
-			var concurrentCalls bool // set to true if we observe concurrent calls
-			var stop bool
-			var mutex sync.Mutex
-			cond := sync.NewCond(&mutex)
+			var concurrentCalls atomic.Value // set to true if we observe concurrent calls
 
+			addCh := make(chan struct{})
 			m.EXPECT().Add(gomock.Any()).DoAndReturn(func(args ...interface{}) error {
-				cond.L.Lock()
-				defer cond.L.Unlock()
-				counter += 1
-				callIdx += 1
-				if callIdx == 1 { // first call
-					// we wait until we need to stop
-					// we stop when one of these conditions is met:
-					// 1) we detected concurrent Add calls (test is a success), or
-					// 2) the test timed out
-					for !stop {
-						cond.Wait()
-					}
-				} else if counter > 1 {
-					concurrentCalls = true
-					stop = true
-					cond.Signal()
+				select {
+				case <-addCh:
+					concurrentCalls.Store(true)
+				case addCh <- struct{}{}:
 				}
 				return nil
 			}).AnyTimes()
 
-			wg.Add(2)
-			go func() {
-				tc.installFn(ofClient, tc.cacheKey)
-				wg.Done()
-			}()
-			go func() {
-				tc.installFn(ofClient, tc.cacheKey)
-				wg.Done()
-			}()
-
-			// enforce a timeout of one second for the test
-			ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+			var wg sync.WaitGroup
+			done := make(chan struct{})
+
+			for i := 0; i < 2; i++ {
+				wg.Add(1)
+				cacheKey := fmt.Sprintf(tc.cacheKeyFormat, i)
+				go func() {
+					defer wg.Done()
+					tc.fn(ofClient, cacheKey)
+				}()
+			}
 			go func() {
-				<-ctx.Done()
-				cond.L.Lock()
-				defer cond.L.Unlock()
-				stop = true
+				defer close(done)
+				wg.Wait()
 			}()
 
-			wg.Wait()
-			cancel()
-
-			assert.True(t, concurrentCalls)
+			select {
+			case <-time.After(time.Second):
+				t.Fatal("timeout, maybe there are some deadlocks")
```

Comment:
should we cleanup the other 3 goroutines in this case?

Reply:
Thanks, enhanced.